### PR TITLE
Add forward traffic to another HTTP proxy for the server (only for LocalProtocol::Tcp)

### PR DESCRIPTION
**Pull request by CorrM**
That help #324

Add this arguments to server CLI:
```
-p, --http-proxy <USER:PASS@HOST:PORT>
          If set, will use this http proxy to connect to the server [env: HTTP_PROXY=]
      --http-proxy-login <LOGIN>
          If set, will use this login to connect to the http proxy. Override the one from --http-proxy [env: WSTUNNEL_HTTP_PROXY_LOGIN=]
      --http-proxy-password <PASSWORD>
          If set, will use this password to connect to the http proxy. Override the one from --http-proxy [env: WSTUNNEL_HTTP_PROXY_PASSWORD=]
```

### CODE REVIEW

**Comment by CorrM** [comment]
@erebe

File: src/tunnel/server.rs [hunk 1]
```diff
@@ -58,28 +58,32 @@ async fn run_tunnel(
...
                 server_config.socket_so_mark,
                 timeout.unwrap_or(Duration::from_secs(10)),
                 &server_config.dns_resolver,
-            )
-            .connect(&None)
-            .await?;
+            );
+            let (rx, tx) = match &server_config.http_proxy {
+                None => connector.connect(&None).await?,
+                Some(proxy_url) => connector.connect_with_http_proxy(proxy_url, &None).await?,
```
Comment (erebe) [inline src/tunnel/server.rs:70]: You cant forward udp traffic with an HTTP proxy. It is going to work only for TCP based protocol.

File: src/tunnel/server.rs [hunk 2]
```diff
@@ -100,6 +104,10 @@ async fn run_tunnel(
             static SERVERS: Lazy<Mutex<HashMap<(Host<String>, u16), mpsc::Receiver<Item>>>> =
                 Lazy::new(|| Mutex::new(HashMap::with_capacity(0)));
 
+            if server_config.http_proxy.is_some() {
```
Comment (erebe) [inline src/tunnel/server.rs:107]: You can remove the error check. Reverse tunneling is not going to use the HTTP proxy as, it is incoming traffic to the server and after is going throught the tunnel

File: src/tunnel/server.rs [hunk 3]
```diff
@@ -117,6 +125,10 @@ async fn run_tunnel(
...
```
Comment (erebe) [inline src/tunnel/server.rs:128]: same

File: src/tunnel/server.rs [hunk 4]
```diff
@@ -133,6 +145,10 @@ async fn run_tunnel(
...
```
Comment (erebe) [inline src/tunnel/server.rs:148]: same

File: src/tunnel/server.rs [hunk 5]
```diff
@@ -150,6 +166,10 @@ async fn run_tunnel(
...
```
Comment (erebe) [inline src/tunnel/server.rs:169]: same

File: src/tunnel/server.rs [hunk 6]
```diff
@@ -169,6 +189,10 @@ async fn run_tunnel(
...
```
Comment (erebe) [inline src/tunnel/server.rs:192]: same

**Comment by erebe** [comment]
Almost, just a few stuff to remove, but looks ok overall.

**Comment by erebe** [comment]
I have a question as I am not really sure of the answer. 

What should be for you the behavior of the server if a proxy is configured but the requested traffic cannot use the proxy. (i.e: udp). Should it forward the traffic or return an error that it can't honor the proxy set ?

**Comment by CorrM** [comment]
> I have a question as I am not really sure of the answer.
> 
> What should be for you the behavior of the server if a proxy is configured but the requested traffic cannot use the proxy. (i.e: udp). Should it forward the traffic or return an error that it can't honor the proxy set ?

Should be an error, As for me when I use a proxy I want to use it to bypass something or act as I live in another country,  if anything happens to the proxy(can't handle the request for any reason) I want to lose the connection instead of expose my real IP

**Comment by erebe** [comment]
Alright, thank you for the PR :)

I am planning to make a release in the following weeks, I have some refactoring work I want to do before the next one. And also do some testing to make sure everything is right !

**Comment by CorrM** [comment]
> Alright, thank you for the PR :)
> 
> I am planning to make a release in the following weeks, I have some refactoring work I want to do before the next one. And also do some testing to make sure everything is right !

BTW new project file struct are much better, keep good work <3

**Pull request merged**